tensorflow/compiler/xla/service/algebraic_simplifier.cc

```diff
@@ -1665,6 +1668,228 @@ StatusOr<HloInstruction*> AlgebraicSimplifierVisitor::OptimizeDotOfGather(
+  CHECK(lhs_contracting_dims.size() == 1);
+  const std::vector<std::pair<int64, int64>>& unmodified_dims =
+      ShapeUtil::DimensionsUnmodifiedByReshape(lhs->operand(0)->shape(),
+                                               lhs->shape());
+  if (unmodified_dims.size() != lhs->shape().rank() - 1) {
+    return nullptr;
+  }
+  for (const auto &pair : unmodified_dims) {
+    if (pair.second == lhs_contracting_dims[0]) {
+      // If the contracting dim is unmodified, and only one dim is modified,
+      // then that dim must be a non-contracting dim.
+      return nullptr;
+    }
+  }
+  // Virtually pull the reshape into the dot. Now the dot is equivalent to a
+  // new dot with "unsquished" lhs contracting dims. We don't need to actually
+  // create a new dot instruction. We can just keep track of lhs and
+  // lhs_contracting_dims.
+  CHECK(lhs->operand(0)->shape().rank() > lhs->shape().rank());
+  for (int64 i = 1; i <= lhs->operand(0)->shape().rank() - lhs->shape().rank();
+       ++i) {
+    lhs_contracting_dims.Add(lhs_contracting_dims[0] + i);
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // Check if lhs_transpose only permutes the contracting dims.
+  const auto& transpose_dims = lhs->dimensions();
+  for (int64 i = 0; i < transpose_dims.size(); ++i) {
+    if (transpose_dims[i] != i &&
+        !(absl::c_linear_search(lhs_contracting_dims, i) &&
+          absl::c_linear_search(lhs_contracting_dims, transpose_dims[i]))) {
+      return nullptr;
+    }
+  }
+  // Virtually pull the transpose into the dot. Now the dot is equivalent to
+  // a new dot with new lhs transpose dims equal to the transpose dims within
+  // the old lhs contracting dims.
+  auto old_lhs_contracting_dims = lhs_contracting_dims;
+  lhs_contracting_dims.Clear();
+  for (const auto dim : old_lhs_contracting_dims) {
+    lhs_contracting_dims.Add(lhs->dimensions(dim));
+  }
+  lhs = lhs->mutable_operand(0);
+
+  // All checks are passed at this point.
+  //
+  // Transform lhs. Remove the transpose and reshape by sorting the lhs
+  // contracting dims and squish them into a single one. We don't actually
+  // squish the lhs_contracting_dims here because we still need the unsquished
+  // contracting dims to invert reshape and transpose.
+  absl::c_sort(lhs_contracting_dims);
+  lhs = computation_->AddInstruction(
+      HloInstruction::CreateReshape(reshape->shape(), lhs));
+
+  // Transform rhs.
+  // Let's walk through an example to see how rhs transformation is done and the
+  // meaning of some variables used in the implementation. Say we have the
+  // following original Hlo IR:
+  //
+  //   t0 = F32[2, 2, 3] parameter(0)
+  //   t1 = F32[2, 3, 2] transpose(t0) dimensions={0, 2, 1}
+  //   t2 = F32[2, 6] reshape(t1)
+  //   t3 = F32[6, 2] constant(...)
+  //   dot = F32[2, 2] dot(t2, t3) lhs_contracting_dims={1},
+  //                               rhs_contracting_dims={0}
+  //
+  // 1) The sorted lhs_contracting_dims is {1, 2}. We compute t1's subshape
```

Comment:
Perhaps clearer:

```
At this point in the function, we have decided that the second and third dims of t0 can be switched, and we have "virtually decomposed" the program to:

  t0 = f32[2,2,3] parameter(0)
  t2' = f32[2,6] reshape(t1)
  t3' = f32[6,2] ???
  dot = f32[2,2] dot(t2', t3') lhs_contracting_dims={1}, rhs_contracting_dims={0}

All that remains is to fill in the value of t3'.  To do this, we have to unsquish the contracting dimensions in t3 and then apply the inverse of the transpose from t1.
```

I think much or all of the remainder of the comment isn't necessary because I believe the reshape-inversion code can be simplified, see below.